[PLUGIN-632] - Added: ServiceAccountType as JSON, unit tests, validations

flakrimjusufi opened the pull request:
Make Google drive/sheet plugins production ready
JIRA Ticket: https://cdap.atlassian.net/browse/PLUGIN-632

flakrimjusufi committed:
[PLUGIN-632] - Added: ServiceAccountType as JSON, more tests, validations, documentation and re-factoring of GoogleDrive and GoogleSheets plugin

jasir99 commented:
lgtm

rmstar commented on src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormat.java:
```diff
@@ -18,6 +18,9 @@
 
 import com.github.rholder.retry.RetryException;
 import com.google.api.services.drive.model.File;
+import com.google.gson.JsonObject;
```
Why are these imports added?

flakrimjusufi replied:
These imports were left behind after we did some additional tests. Thanks for letting me know. I removed them.

rmstar commented on src/main/java/io/cdap/plugin/google/drive/source/SchemaBuilder.java:
```diff
@@ -27,7 +27,7 @@
  */
 public class SchemaBuilder {
-  public static final String SCHEMA_ROOT_RECORD_NAME = "FileFromFolder";
+  public static final String SCHEMA_ROOT_RECORD_NAME = "etlSchemaBody";
```
why was this change needed?

flakrimjusufi replied:
The schema was not being updated in _Google Drive Source_ and initially we thought that this was the case, but then we found out that the issue was related with widget, not with `SCHEMA_ROOT_RECORD_NAME`. Reverted to its initial name in the updated PR.

rmstar commented on src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java:
```diff
@@ -98,7 +105,7 @@ protected Credential getCredentials() throws IOException {
 
   protected List<String> getRequiredScopes() {
-    return Collections.singletonList(DriveScopes.DRIVE_READONLY);
+    return Collections.singletonList(DriveScopes.DRIVE);
```
why did we need to make this change?

flakrimjusufi replied:
This was also as part of the testings that we have done. We were dealing with this error message when testing the pipeline:

`404 Not Found { "code" : 404, "errors" : [ { "domain" : "global", "location" : "fileId", "locationType" : "parameter", "message" : "File not found: 1ZqAr0q6gITvjqvfOJqkSIxd-usTNhvJy.", "reason" : "notFound" } ], "message" : "File not found: 1ZqAr0q6gITvjqvfOJqkSIxd-usTNhvJy." }. Please check the system logs for more details.
`

...and checking the documentation of Google Drive API for 404 error type, stating that:

To fix this error:

```
1. Inform the user they don't have read access to the file or that the file doesn't exist.
2. Instruct the user to ask the owner for permission to the file.
```
...we changed the SCOPE from `DRIVE_READONLY` to `DRIVE`.

But afterwards, when we verified that in fact we were not in the right path when testing the pipelines and DRIVE_READONLY was enough for read/write, we changed back to its initial SCOPE. 

In the last commit, we reverted back the scope to `DRIVE_READONLY`.

rmstar replied:
With SCOPE set to `DRIVE` does everything work correctly?

flakrimjusufi replied:
Actually, everything also works correctly with SCOPE set to `DRIVE_READONLY`, that's why we reverted back the code.

rmstar commented on src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:
```diff
@@ -158,25 +176,30 @@ private boolean validateOAuth2Properties(FailureCollector collector) {
-  private boolean validateAccountFilePath(FailureCollector collector) {
-    if (!containsMacro(ACCOUNT_FILE_PATH)) {
+  private boolean validateServiceAccount(FailureCollector collector) {
+    if (isServiceAccountFilePath() != null && isServiceAccountFilePath()) {
```
store the return value in a variable so you don't have to call `isServiceAccountFilePath()` twice. Same thing for `isServiceAccountJson()` below.

flakrimjusufi replied:
PR updated.

rmstar reviewed (commented):
Please test source and sink plugins in the following scenarios:
1. In CDF with file path set to auto-detect.
2. In CDF, specify file path
3. In CDF, specify Service Account JSON
4. In CDAP, specify file path
5. In CDAP, specify Service Account JSON
6. Test with service account type, file path/JSON set as macro param.

flakrimjusufi commented:
> Please test source and sink plugins in the following scenarios:
> 
> 1. In CDF with file path set to auto-detect.
> 2. In CDF, specify file path
> 3. In CDF, specify Service Account JSON
> 4. In CDAP, specify file path
> 5. In CDAP, specify Service Account JSON
> 6. Test with service account type, file path/JSON set as macro param.

1. I'm unable to test with file path set to auto-detect in CDF because my keys have insufficient permissions. To be more specific, I've used `flakrim@cirus.co` to test the plugin since I don't have access to Google Drive with my partner account `flakrim@88547414.corp-partner.google.com`. In CDF, we only have access with partnership accounts. 

2. Is there a way to test the pipelines in CDF while specifying the file path? As far as I know, in CDF we can only test with Service Account as JSON. Please correct me if my assumption is wrong. 

3. Test passed successfully. 

4. Test passed successfully. 

5. Test passed successfully. 

6. There's an issue when you test with MACRO. We were getting this error message:

`Java.util.concurrent.ExecutionException: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 5965
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:294) ~[com.google.guava.guava-13.0.1.jar:na]
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:281) ~[com.google.guava.guava-13.0.1.jar:na]
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116) ~[com.google.guava.guava-13.0.1.jar:na]
`
That's because of this block of code: 

https://github.com/data-integrations/google-drive/blob/a9be71186e7aae818c5b5c78a5d33e254c1f5368/src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormatProvider.java#L40-L43

After debugging this issue, we found out that the entire config was serialized. 

Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.

However, we can proceed with the deserialization of config and we can include Macros in this PR if needed.

CuriousVini commented:
@flakrimjusufi
Do you know what permissions are missing from your account?

CuriousVini commented:
Also are you noticing permissions issues at pipeline runtime? If so, could you please make sure dataproc service account (GCE default service account) has permissions to access drive directories/files.

rmstar commented:
> Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.

Macros are supported, so we should make sure they work correctly. Please fix it as part of this PR.

flakrimjusufi commented:
> @flakrimjusufi
> Do you know what permissions are missing from your account?

With our partnership accounts, we don't have access in any of gmail, google drive, etc. So, to test Google Drive Source/Sink in data proc, I need one of the followings:  

- My account in cirus (which I've used to create a project and work with Google Drive API) to have access in data proc. 
- My partnership account to have access in Google Drive API. 

Permission issues are showing up in validation when file path is set to `auto-detect` in datafusion, I can't run the pipeline.

flakrimjusufi commented:
> > Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.
> 
> Macros are supported, so we should make sure they work correctly. Please fix it as part of this PR.

@rmstar 

In the last update, we deserialized the config and added Macros.

rmstar commented on src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:
```diff
@@ -238,4 +298,53 @@ public String getClientSecret() {
+
+  @Nullable
+  public String getServiceAccountFilePath() {
+    if (containsMacro(ACCOUNT_FILE_PATH)  || accountFilePath == null ||
```
nit: replace accountFilePath == null || accountFilePath.isEmpty() -> Strings.insNullOrEmpty(accountFilePath)

rmstar commented on src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:
```diff
@@ -238,4 +298,53 @@ public String getClientSecret() {
+  }
+
+  @Nullable
+  public String getServiceAccount() {
```
This is only used in GoogleBasicAuthConfig.java where you're checking if the returned string is null or empty. The string value is unused. 

It seems this is unnecessary. You just need something like this:
```
if (config.isServiceAccountFilePath()) {
...
} else if (config. isServiceAccountJson()) {
...
} else {
// use default credentials
}
```

flakrimjusufi replied:
Removed in the last update.

rmstar commented on src/test/java/io/cdap/plugin/google/common/GoogleAuthBaseConfigTest.java:
```diff
@@ -0,0 +1,45 @@
+import org.junit.Test;
+
+public class GoogleAuthBaseConfigTest {
+
```
can you also add tests for validation error handling?

flakrimjusufi replied:
Added tests for validation error handling in the last update.

flakrimjusufi commented:
@rmstar 

In the last commit, GoogleSheets plugins are also updated with serviceAccountType as JSON, macros and documentation. 

Both Source and Sink were tested thoroughly and everything is working well.

rmstar commented on src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:
```diff
@@ -47,6 +51,11 @@
+  public static final String NAME_SERVICE_ACCOUNT_JSON = "serviceAccountJSON";
+  public static final String SERVICE_ACCOUNT_FILE_PATH = "filePath";
+  public static final String SERVICE_ACCOUNT_JSON = "JSON";
+  public static final String SCHEMA = "schema";
```
where is this used?

flakrimjusufi replied:
`SCHEMA` is used in deserialization of config: 

https://github.com/data-integrations/google-drive/blob/3074341092eecd2522ea5ce9ac1c40877e0fc03f/src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveSourceConfig.java#L358-L360

rmstar commented on src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:
```diff
@@ -228,16 +230,47 @@
   public Schema getSchema() {
-    if (schema == null) {
-      if (dataSchemaInfo.isEmpty()) {
-        throw new RuntimeException("There are no headers to process. " +
```
Is this not relevant anymore?

flakrimjusufi replied:
It doesn't look relevant to me because we are already checking if Source folder (directory) contains any spreadsheets files in here: 

https://github.com/data-integrations/google-drive/blob/3074341092eecd2522ea5ce9ac1c40877e0fc03f/src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java#L348-L354

...but I will revert this RuntimeException in getSchema().

rmstar replied:
hm why are we throwing a RuntimeException here? If config validation fails, we should add the failure to the failure collector (pass in FailureCollector, call collector.addFailure()).

flakrimjusufi replied:
PR updated.

rmstar commented on src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:
```diff
@@ -228,16 +230,47 @@
-        throw new RuntimeException("There are no headers to process. " +
-          "Perhaps no validation step was executed before schema generation.");
-      }
+    if (schema == null && !containsMacro("serviceAccountType")) {
```
What happens if service account type is not macro (set to default, i.e. file path), but the file path is a macro?

flakrimjusufi replied:
Thanks for catching this. In this scenario the validation would fail. In the next update, I will also check if file path or service acccount json contains Macro.

rmstar reviewed (commented):
Can you add an ETL test with macros for google drive and sheets?

flakrimjusufi commented:
> Can you add an ETL test with macros for google drive and sheets?

I'm seeing lots of dependencies that are missing in `pom.xml`. Should I create a separate PR only with ETL Tests? It will take some time because first I will need to include all the missing dependencies, create the ETL Tests and then I will have to do a thorough test of both Google Drive and Google Sheets plugin.

rmstar commented:
> Should I create a separate PR only with ETL Tests?

You can add a test in a separate PR if that's easier for you.

rmstar reviewed (approved):
please squash commits

flakrimjusufi commented:
> please squash commits

Commits squashed.
